README.md
# Editable

⚡️A highly customizable, editable table package for Flutter projects.

### Specs
[![pub](https://img.shields.io/pub/v/editable.svg?style=flat)](https://pub.dev/packages/editable)

This package allows you to create editable tables and spreadsheets with ease, either by providing inital row and column count
to display an empty table, or use it with predefined rows and column data sets.

**It is compatible with Flutter Desktop (Windows, Linux, MacOS), IOS, Android, and Web**. ❤️

## Features
* Save Edited Rows
* Save Edited cell 
* Customize cell data
* Add Empty rows
* Create Empty tables
* Customizable Save button
* Customize Table headers
* Zebra - Stripe table
* Style table border
* Adjustable Column widths


### Spread Some :heart:
[![GitHub followers](https://img.shields.io/github/followers/godilite.svg?style=social&label=Follow)](https://github.com/godilite)

## Support us

You can buy us a cup of coffee. Sponsor our next milestones

[![Patreon](https://c5.patreon.com/external/logo/become_a_patron_button.png)](https://www.patreon.com/godilite)


## Getting started

In the `pubspec.yaml` of your flutter project, add the following
dependency:

```yaml
dependencies:
  ...
  editable: "^1.1.4"
```

In your library add the following import:

```dart
import 'package:editable/editable.dart';
```

## API

### Create Table

To create a new table, use the `Editable()` widget class and provide the table data  properties as follows:
- `column`: an array of objects example
     example:
     ```dart 
     List cols = [
       {"title":'Name', 'widthFactor': 0.2, 'key':'name'},
       {"title":'Date', 'widthFactor': 0.1, 'key':'date'},
       {"title":'Month', 'widthFactor': 0.1, 'key':'month'},
       {"title":'Status',  'key':'status'},
     ]; 
     ```
     [title] is the column heading
    
     [widthFactor] a custom size ratio of each column width, if not provided, defaults to  [columnRatio = 0.20]
      ```dart
      'widthFactor': 0.1 //gives 10% of screen size to the column
      'widthFactor': 0.2 //gives 20% of screen size to the column
      ```
    
     [key] an identifyer preferably without space and special characters

- `rows`: an array of objects, where each object key represents a column [key],
  example:
    ```dart
    List rows = [
        {"name": 'James Joe', "date":'23/09/2020',"month":'June',"status":'completed'},
        {"date":'12/4/2020',"month":'March',"name": 'Daniel Paul', "status":'new'},
      ];
    ```
  Each objects *DO NOT* have to be positioned in same order as its column
 
- `rowCount`: Interger value of number of rows to be generated: Optional if row data is provided
- `columnCount`: Interger value of number of columns to be generated: Optional if column data is provided 
- `columnRatio`: fraction of the screen width occupied by each column. This is usefull for shrinking and expanding table width
  example: 
  ```dart
    columnRation: 0.2 //sets each column to occupy 20 percent of screen width
  ```

### Customization
  - `borderColor`:  Color of table border
  - `borderWidth`: width of table borders  
  - `tdPaddingLeft`:  Table data cell padding left 
  - `tdPaddingTop`: Table data cell padding top
  - `tdPaddingRight`:  Table data cell padding right;
  - `tdPaddingBottom`: Table data cell padding bottom;
  - `tdAlignment`: Aligns the table data ;
  - `tdStyle`: Style the table data;
  - `thPaddingLeft`: Table header cell padding left;
  - `thPaddingTop`: Table header cell padding top;
  - `thPaddingRight`:  Table header cell padding right;  
  - `thPaddingBottom`: Table header cell padding bottom;
  - `trHeight`: Table Row Height, cannot be less than 40.0 ;
  - `thWeight`: Table headers fontweight;
  - `thSize`: Table headers fontSize;
  - `tdEditableMaxLines` : Max lines allowed in editable text, default: 1 (longer data will not wrap and be hidden), setting to 100 will allow wrapping and not increase row size;
  - `showSaveIcon`:  Toogles the save button, if [true] displays an icon to save rows, 
        adds an addition column to the right
  - `saveIcon`: Icon for to save row data, example:
        ```dart
        saveIcon : Icons.add
        ```
  - `saveIconColor`: Color for the save Icon
  - `saveIconSize`:   Size for the saveIcon
  - `showCreateButton`:  displays a button that adds a new row onPressed
  - `stripeColor1`: The first row alternate color, if stripe is set to true
  - `stripeColor2`:     The Second row alternate color, if stripe is set to true;
  
   - `zebraStripe`: Enable zebra-striping, set to false by default
      ```// if zebraStripe is enabled, you can style the colors [stripeColor1] and [stripeColor2]```
  - `createButtonAlign`: Aligns the button for adding new rows;
  - `createButtonIcon`: Icon displayed in the create new row button;
  - `createButtonColor`: Color for the create new row button;   
  - `createButtonShape`:   border shape of the create new row button
        ```
        createButtonShape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8)
        )
        ```
  - `createButtonLabel`:  Label for the create new row button
 
### Save methods

- `onSubmitted`:  [onSubmitted] callback is triggered when the enter button is tapped (for desktop users)
     Or when edit is complete(on mobile) on a table data cell
    it returns a value of the edited cell data
 
- `onRowSaved`: [onRowSaved] callback is triggered when a [saveButton] is pressed.
   returns only values if row is edited, otherwise returns a string ['no edit']

- `EditableState` key: To get all Edited Rows outside the current context, you can access the editable state using a GlobalKey,
  example:
  ```dart
       /// Create a Key for EditableState
      final _editableKey = GlobalKey<EditableState>(); 

    
      /// Function to add a new row
      /// Using the global key assigined to Editable widget
      /// Access the current state of Editable
      void _addNewRow() {
        setState(() {
          _editableKey.currentState.createRow();
        });
      }

      ///Print only edited rows.
      void _printEditedRows() {
        List editedRows = _editableKey.currentState.editedRows;
        print(editedRows);
      }

        @override
      Widget build(BuildContext context) {
        return Scaffold(
          appBar: AppBar(
            leadingWidth: 200,
            leading: FlatButton.icon(
                onPressed: () => _addNewRow(),
                icon: Icon(Icons.add),
                label: Text(
                  'Add',
                  style: TextStyle(fontWeight: FontWeight.bold),
                )),
            title: Text(widget.title),
            actions: [
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: FlatButton(
                    onPressed: () => _printEditedRows(),
                    child: Text('Print Edited Rows',
                        style: TextStyle(fontWeight: FontWeight.bold))),
              )
            ],
          ),
          body: Editable(
            key: _editableKey, //Assign Key to Widget
            columns: cols,
            rows: rows,
            zebraStripe: true,
            stripeColor2: Colors.grey[200],
            borderColor: Colors.blueGrey,
          ),
        );
      }

  ```
# Screenshots

Editable Table with Data (zebraStripe)  |Empty SpreadSheet (zebraStripe)    | Empty SpreadSheet (No Stripe)
:-------------:|:-------------:|:-------------:
<img src="https://user-images.githubusercontent.com/41484542/96218374-85cb8100-0f7c-11eb-9bd3-f3154073d747.png"/> | <img src="https://user-images.githubusercontent.com/41484542/96218512-caefb300-0f7c-11eb-93e1-77fd66f7cc2b.png"/> | <img src="https://user-images.githubusercontent.com/41484542/96218567-e490fa80-0f7c-11eb-8df1-03aebd4b4ee4.png"/>

Table with Save Icon
:-------------:|
<img src="https://user-images.githubusercontent.com/41484542/95015531-4da37480-0645-11eb-965a-f3d3507c18da.png"/>



# Tutorials
The Following are resources to help you build with Editable: 
- [Creating Editable Tables With Flutter](https://medium.com/swlh/creating-editable-tables-with-flutter-5536b593c29a)
- [Keeping it local: Managing a Flutter app's data (Andrew Fitz Gibbon)](https://www.youtube.com/watch?v=xjhxm3foCMk&t=2743s)


# Contribution

I highly encourage the community to step forward and improve this
library further. You can fix any reported bug, propose or implement new
features, write tests, etc.

Here is a quick list of things to remember
* Check the open issues before creating a new one,
* Help me in reducing the number of open issues by fixing any existing
  bugs,
* Check the roadmap to see if you can help in implementing any new
  feature,
* You can contribute by writing unit and integration tests for this
  library,
* If you have any new idea that aligns with the goal of this library,
  feel free to raise a feature request and discuss it.

# Author

This Editable table package for Flutter is developed by [Godwin Asuquo](https://github.com/godilite)

# Contributors
[CY Uket](https://github.com/cyuket)
[AmitB](https://github.com/Amitbhave)
[Tim Maffet](https://github.com/timmaffett)
[Thumbert](https://github.com/thumbert)
❤️Thank you all for your contributions

Also, as always, please give us a star to help!




CHANGELOG.md
## [1.0.0] - 04/10/2020.

* This Initial release has the following features:
* Save Edited Rows
* Save edited cell 
* Customize cell data
* Add empty rows
* Create empty tables
* Customizable save button
* Customize table headers

## [1.0.1] - 04/10/2020.

* minor edit to package metatags

## [1.0.2] - 04/10/2020.

* minor edit to package readme

## [1.0.3] - 04/10/2020.

* minor edit to package readme
* compatability check passed for all OS

## [1.0.4] - 04/10/2020.

* fix: removes Space showing on small screen sizes if saveIcon is shown
* compatability check passed for all OS
* formats code based on dartfmt

## [1.0.5] - 05/10/2020.

* Bug Fix: table header alignments when saveRow icon is shown

## [1.0.6] - 07/10/2020.

* Bug Fix: Align table headers
* Refactor: Adds helper functions
* Change default button for createRowButton

## [1.0.7] - 08/10/2020.

* Bug Fix: Table color set to transparent
* This allows table to take parent widget color property

## [1.0.8] - 11/10/2020.

* New Feature: Horizontal and vertical scrollable
* New Table property: [columnRatio] sets the fraction of screen width occupied by each column

## [1.0.9] - 12/10/2020.

* Bug fix: A known behaviour where table doesnot rebuild with new data was fixed. This fix ensures that new rows can be added at runtime

## [1.1.0] - 12/10/2020.

* Bug fix: issues with width constraint for column element

## [1.1.1] - 12/10/2020.

* New feature: Zebra stripe table rows
* Set prefered colors for zebra striping
* Order row data to column [key] parameter 

## [1.1.2] - 28/10/2020.

* New feature: Adjustable Column width
* Replaced the [index] parameter of the column with [widthFactor]
* [widthFactor] is a double, used to assign percentage width to each column

## [1.1.3] - 04/11/2020.

* New feature: Use GlobalKey State to get edited row data
* New feature: Use GlobalKey State to create new row
* This enable creating row and saving edited rows anywhere outside the Editable widget

## [1.1.4] - 16/02/2021

* added tdEditableMaxLines parameter to allow editable data to wrap and not be hidden when it overflows (setting to 100 allows editable data to behave same as the NON editable data)
* Added support for specifying max lines for editable data cells so that they can wrap and have all data visible.
* added support for completely controlling the style and alignment of the table headers.

## [2.0.0] - 04/03/2021

* Migrates to sound null-safety
* Remove deprecated example widgets

lib/editable.dart
// Copyright 2020 Godwin Asuquo. All rights reserved.
//
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

library editable;

import 'package:flutter/material.dart';
import 'commons/helpers.dart';
import 'widgets/table_body.dart';
import 'widgets/table_header.dart';

class Editable extends StatefulWidget {
  /// Builds an editable table using predefined row and column counts
  /// Or using a row and header data set provided
  ///
  /// if no data is provided for [row] and [column],
  /// [rowCount] and [columnCount] properties must be set
  /// this will generate an empty table
  ///
  /// it is useful for rendering data from an API or to create a spreadsheet-like
  /// data table
  ///
  /// example:
  ///
  /// ```dart
  ///  Widget build(BuildContext context) {
  ///   return Scaffold(
  ///     body: Column(
  ///       children: <Widget>[
  ///           Expanded(
  ///           flex: 1,
  ///           child: EdiTable(
  ///               showCreateButton: true,
  ///               tdStyle: TextStyle(fontSize: 20),
  ///               showSaveIcon: false,
  ///               borderColor: Colors.lightBlue,
  ///               columnCount: 4,
  ///               rowCount: 8
  ///              ),
  ///           ).
  ///         ]
  ///       ),
  ///   );
  /// }
  /// ```
  Editable(
      {Key? key,
      this.columns,
      this.rows,
      this.columnRatio = 0.20,
      this.onSubmitted,
      this.onRowSaved,
      this.columnCount = 0,
      this.rowCount = 0,
      this.borderColor = Colors.grey,
      this.tdPaddingLeft = 8.0,
      this.tdPaddingTop = 8.0,
      this.tdPaddingRight = 8.0,
      this.tdPaddingBottom = 12.0,
      this.thPaddingLeft = 8.0,
      this.thPaddingTop = 0.0,
      this.thPaddingRight = 8.0,
      this.thPaddingBottom = 0.0,
      this.trHeight = 50.0,
      this.borderWidth = 0.5,
      this.thWeight = FontWeight.w600,
      this.thSize = 18,
      this.showSaveIcon = false,
      this.saveIcon = Icons.save,
      this.saveIconColor = Colors.black12,
      this.saveIconSize = 18,
      this.tdAlignment = TextAlign.start,
      this.tdStyle,
      this.tdEditableMaxLines = 1,
      this.thAlignment = TextAlign.start,
      this.thStyle,
      this.thVertAlignment = CrossAxisAlignment.center,
      this.showCreateButton = false,
      this.createButtonAlign = CrossAxisAlignment.start,
      this.createButtonIcon,
      this.createButtonColor,
      this.createButtonShape,
      this.createButtonLabel,
      this.stripeColor1 = Colors.white,
      this.stripeColor2 = Colors.black12,
      this.zebraStripe = false,
      this.focusedBorder})
      : super(key: key);

  /// A data set to create headers
  ///
  /// Can be null if blank columns are needed, else:
  /// Must be array of objects
  /// with the following keys: [title], [widthFactor] and [key]
  ///
  /// example:
  /// ```dart
  /// List cols = [
  ///   {"title":'Name', 'widthFactor': 0.1, 'key':'name', 'editable': false},
  ///   {"title":'Date', 'widthFactor': 0.2, 'key':'date'},
  ///   {"title":'Month', 'widthFactor': 0.1, 'key':'month', 'editable': false},
  ///   {"title":'Status', 'widthFactor': 0.1, 'key':'status'},
  /// ];
  /// ```
  /// [title] is the column heading
  ///
  /// [widthFactor] a custom size ratio of each column width, if not provided, defaults to  [columnRatio = 0.20]
  /// ```dart
  /// 'widthFactor': 0.1 //gives 10% of screen size to the column
  /// 'widthFactor': 0.2 //gives 20% of screen size to the column
  /// ```
  ///
  /// [key] an identifier preferably a short string
  /// [editable] a boolean, if the column should be editable or not, [true] by default.
  final List? columns;

  /// A data set to create rows
  ///
  /// Can be null if empty rows are needed. else,
  /// Must be array of objects
  /// with keys matching [key] provided in the column array
  ///
  /// example:
  /// ```dart
  ///List rows = [
  ///          {"name": 'James Joe', "date":'23/09/2020',"month":'June',"status":'completed'},
  ///          {"name": 'Daniel Paul', "date":'12/4/2020',"month":'March',"status":'new'},
  ///        ];
  /// ```
  /// each objects DO NOT have to be positioned in same order as its column

  final List? rows;

  /// Interger value of number of rows to be generated:
  ///
  /// Optional if row data is provided
  final int rowCount;

  /// Interger value of number of columns to be generated:
  ///
  /// Optional if column data is provided
  final int columnCount;

  /// aspect ration of each column,
  /// sets the ratio of the screen width occupied by each column
  /// it is set in fraction between 0 to 1.0
  /// 0.8 indicates 80 percent width per column
  final double columnRatio;

  /// Color of table border
  final Color borderColor;

  /// width of table borders
  final double borderWidth;

  /// Table data cell padding left
  final double tdPaddingLeft;

  /// Table data cell padding top
  final double tdPaddingTop;

  /// Table data cell padding right
  final double tdPaddingRight;

  /// Table data cell padding bottom
  final double tdPaddingBottom;

  /// Aligns the table data
  final TextAlign tdAlignment;

  /// Style the table data
  final TextStyle? tdStyle;

  /// Max lines allowed in editable text, default: 1 (longer data will not wrap and be hidden), setting to 100 will allow wrapping and not increase row size
  final int tdEditableMaxLines;

  /// Table header cell padding left
  final double thPaddingLeft;

  /// Table header cell padding top
  final double thPaddingTop;

  /// Table header cell padding right
  final double thPaddingRight;

  /// Table header cell padding bottom
  final double thPaddingBottom;

  /// Aligns the table header
  final TextAlign thAlignment;

  /// Style the table header - use for more control of header style, using this OVERRIDES the thWeight and thSize parameters and those will be ignored.
  final TextStyle? thStyle;

  /// Table headers fontweight (use thStyle for more control of header style)
  final FontWeight thWeight;

  /// Table header label vertical alignment
  final CrossAxisAlignment thVertAlignment;

  /// Table headers fontSize  (use thStyle for more control of header style)
  final double thSize;

  /// Table Row Height
  /// cannot be less than 40.0
  final double trHeight;

  /// Toogles the save button,
  /// if [true] displays an icon to save rows,
  /// adds an addition column to the right
  final bool showSaveIcon;

  /// Icon for to save row data
  /// example:
  ///
  /// ```dart
  /// saveIcon : Icons.add
  /// ````
  final IconData saveIcon;

  /// Color for the save Icon
  final Color saveIconColor;

  /// Size for the saveIcon
  final double saveIconSize;

  /// displays a button that adds a new row onPressed
  final bool showCreateButton;

  /// Aligns the button for adding new rows
  final CrossAxisAlignment createButtonAlign;

  /// Icon displayed in the create new row button
  final Icon? createButtonIcon;

  /// Color for the create new row button
  final Color? createButtonColor;

  /// border shape of the create new row button
  ///
  /// ```dart
  /// createButtonShape: RoundedRectangleBorder(
  ///   borderRadius: BorderRadius.circular(8)
  /// )
  /// ```
  final BoxShape? createButtonShape;

  /// Label for the create new row button
  final Widget? createButtonLabel;

  /// The first row alternate color, if stripe is set to true
  final Color stripeColor1;

  /// The Second row alternate color, if stripe is set to true
  final Color stripeColor2;

  /// enable zebra-striping, set to false by default
  /// if enabled, you can style the colors [stripeColor1] and [stripeColor2]
  final bool zebraStripe;

  final InputBorder? focusedBorder;

  ///[onSubmitted] callback is triggered when the enter button is pressed on a table data cell
  /// it returns a value of the cell data
  final ValueChanged<String>? onSubmitted;

  /// [onRowSaved] callback is triggered when a [saveButton] is pressed.
  /// returns only values if row is edited, otherwise returns a string ['no edit']
  final ValueChanged<dynamic>? onRowSaved;

  @override
  EditableState createState() => EditableState(
      rows: this.rows,
      columns: this.columns,
      rowCount: this.rowCount,
      columnCount: this.columnCount);
}

class EditableState extends State<Editable> {
  List? rows, columns;
  int? columnCount;
  int? rowCount;

  ///Get all edited rows
  List get editedRows => _editedRows;

  ///Create a row after the last row
  createRow() => addOneRow(columns, rows);
  EditableState({this.rows, this.columns, this.columnCount, this.rowCount});

  /// Temporarily holds all edited rows
  List _editedRows = [];

  @override
  Widget build(BuildContext context) {
    /// initial Setup of columns and row, sets count of column and row
    rowCount = rows == null || rows!.isEmpty ? widget.rowCount : rows!.length;
    columnCount =
        columns == null || columns!.isEmpty ? columnCount : columns!.length;
    columns = columns ?? columnBlueprint(columnCount, columns);
    rows = rows ?? rowBlueprint(rowCount!, columns, rows);

    /// Builds saveIcon widget
    Widget _saveIcon(index) {
      return Flexible(
        fit: FlexFit.loose,
        child: Visibility(
          visible: widget.showSaveIcon,
          child: IconButton(
            padding: EdgeInsets.only(right: widget.tdPaddingRight),
            hoverColor: Colors.transparent,
            icon: Icon(
              widget.saveIcon,
              color: widget.saveIconColor,
              size: widget.saveIconSize,
            ),
            onPressed: () {
              int rowIndex = editedRows.indexWhere(
                  (element) => element['row'] == index ? true : false);
              if (rowIndex != -1) {
                widget.onRowSaved!(editedRows[rowIndex]);
              } else {
                widget.onRowSaved!('no edit');
              }
            },
          ),
        ),
      );
    }

    /// Generates table columns
    List<Widget> _tableHeaders() {
      return List<Widget>.generate(columnCount! + 1, (index) {
        return columnCount! + 1 == (index + 1)
            ? iconColumn(widget.showSaveIcon, widget.thPaddingTop,
                widget.thPaddingBottom)
            : THeader(
                widthRatio: columns![index]['widthFactor'] != null
                    ? columns![index]['widthFactor'].toDouble()
                    : widget.columnRatio,
                thAlignment: widget.thAlignment,
                thStyle: widget.thStyle,
                thPaddingLeft: widget.thPaddingLeft,
                thPaddingTop: widget.thPaddingTop,
                thPaddingBottom: widget.thPaddingBottom,
                thPaddingRight: widget.thPaddingRight,
                headers: columns,
                thWeight: widget.thWeight,
                thSize: widget.thSize,
                index: index);
      });
    }

    /// Generates table rows
    List<Widget> _tableRows() {
      return List<Widget>.generate(rowCount!, (index) {
        return Row(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: List.generate(columnCount! + 1, (rowIndex) {
            var ckeys = [];
            var cwidths = [];
            var ceditable = <bool>[];
            columns!.forEach((e) {
              ckeys.add(e['key']);
              cwidths.add(e['widthFactor'] ?? widget.columnRatio);
              ceditable.add(e['editable'] ?? true);
            });
            var list = rows![index];
            return columnCount! + 1 == (rowIndex + 1)
                ? _saveIcon(index)
                : RowBuilder(
                    index: index,
                    col: ckeys[rowIndex],
                    trHeight: widget.trHeight,
                    borderColor: widget.borderColor,
                    borderWidth: widget.borderWidth,
                    cellData: list[ckeys[rowIndex]],
                    tdPaddingLeft: widget.tdPaddingLeft,
                    tdPaddingTop: widget.tdPaddingTop,
                    tdPaddingBottom: widget.tdPaddingBottom,
                    tdPaddingRight: widget.tdPaddingRight,
                    tdAlignment: widget.tdAlignment,
                    tdStyle: widget.tdStyle,
                    tdEditableMaxLines: widget.tdEditableMaxLines,
                    onSubmitted: widget.onSubmitted,
                    widthRatio: cwidths[rowIndex].toDouble(),
                    isEditable: ceditable[rowIndex],
                    zebraStripe: widget.zebraStripe,
                    focusedBorder: widget.focusedBorder,
                    stripeColor1: widget.stripeColor1,
                    stripeColor2: widget.stripeColor2,
                    onChanged: (value) {
                      ///checks if row has been edited previously
                      var result = editedRows.indexWhere((element) {
                        return element['row'] != index ? false : true;
                      });

                      ///adds a new edited data to a temporary holder
                      if (result != -1) {
                        editedRows[result][ckeys[rowIndex]] = value;
                      } else {
                        var temp = {};
                        temp['row'] = index;
                        temp[ckeys[rowIndex]] = value;
                        editedRows.add(temp);
                      }
                    },
                  );
          }),
        );
      });
    }

    return Material(
      color: Colors.transparent,
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child:
              Column(crossAxisAlignment: widget.createButtonAlign, children: [
            //Table Header
            createButton(),
            Container(
              padding: EdgeInsets.only(bottom: widget.thPaddingBottom),
              decoration: BoxDecoration(
                  border: Border(
                      bottom: BorderSide(
                          color: widget.borderColor,
                          width: widget.borderWidth))),
              child: Row(
                  crossAxisAlignment: widget.thVertAlignment,
                  mainAxisSize: MainAxisSize.min,
                  children: _tableHeaders()),
            ),

            Expanded(
              child: SingleChildScrollView(
                child: Column(
                  children: _tableRows(),
                ),
              ),
            )
          ]),
        ),
      ),
    );
  }

  /// Button for creating a new empty row
  Widget createButton() {
    return Visibility(
      visible: widget.showCreateButton,
      child: Padding(
        padding: EdgeInsets.only(left: 4.0, bottom: 4),
        child: InkWell(
          onTap: () {
            rows = addOneRow(columns, rows);
            rowCount = rowCount! + 1;
            setState(() {});
          },
          child: Container(
            padding: EdgeInsets.all(2),
            decoration: BoxDecoration(
              color: widget.createButtonColor ?? Colors.white,
              boxShadow: [
                BoxShadow(blurRadius: 2, color: Colors.grey.shade400)
              ],
              borderRadius: BorderRadius.circular(10),
              shape: BoxShape.rectangle,
            ),
            child: widget.createButtonIcon ?? Icon(Icons.add),
          ),
        ),
      ),
    );
  }
}
